File: kafka-connect-common/src/test/scala/com/datamountaineer/streamreactor/connect/sink/StructFieldsStringKeyBuilderTest.scala
package com.datamountaineer.streamreactor.connect.sink

import org.apache.kafka.connect.data.{Schema, SchemaBuilder, Struct}
import org.apache.kafka.connect.sink.SinkRecord
import org.scalatest.{Matchers, WordSpec}


class StructFieldsStringKeyBuilderTest extends WordSpec with Matchers {
  "StructFieldsStringKeyBuilder" should {
    "raise an exception if the field is not present in the struct" in {
      intercept[IllegalArgumentException] {
        val schema = SchemaBuilder.struct().name("com.example.Person")
          .field("firstName", Schema.STRING_SCHEMA)
          .field("age", Schema.INT32_SCHEMA)
          .field("threshold", Schema.OPTIONAL_FLOAT64_SCHEMA).build()

        val struct = new Struct(schema).put("firstName", "Alex").put("age", 30)

        val sinkRecord = new SinkRecord("sometopic", 1, null, null, schema, struct, 1)
        StructFieldsStringKeyBuilder(Seq("threshold")).build(sinkRecord)
      }
    }

    "create the row key based on one single field in the struct" in {
      val schema = SchemaBuilder.struct().name("com.example.Person")
        .field("firstName", Schema.STRING_SCHEMA)
        .field("age", Schema.INT32_SCHEMA)
        .field("threshold", Schema.OPTIONAL_FLOAT64_SCHEMA).build()

      val struct = new Struct(schema).put("firstName", "Alex").put("age", 30)

      val sinkRecord = new SinkRecord("sometopic", 1, null, null, schema, struct, 1)
      StructFieldsStringKeyBuilder(Seq("firstName")).build(sinkRecord) shouldBe "Alex"
    }

    "create the row key based on more thant one field in the struct" in {
      val schema = SchemaBuilder.struct().name("com.example.Person")
        .field("firstName", Schema.STRING_SCHEMA)
        .field("age", Schema.INT32_SCHEMA)
        .field("threshold", Schema.OPTIONAL_FLOAT64_SCHEMA).build()

      val struct = new Struct(schema).put("firstName", "Alex").put("age", 30)

      val sinkRecord = new SinkRecord("sometopic", 1, null, null, schema, struct, 1)
      StructFieldsStringKeyBuilder(Seq("firstName", "age")).build(sinkRecord) shouldBe "Alex.30"
    }
  }
}

File: kafka-connect-common/src/test/java/com/datamountaineer/streamreactor/connect/sink/StructFieldsStringKeyBuilderTest.java
package com.datamountaineer.streamreactor.connect.sink;

import org.apache.kafka.connect.data.Schema;
import org.apache.kafka.connect.data.SchemaBuilder;
import org.apache.kafka.connect.data.Struct;
import org.apache.kafka.connect.sink.SinkRecord;
import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class StructFieldsStringKeyBuilderTest {
    @Test(expected = IllegalArgumentException.class)
    public void raiseExceptionIfFieldIsNotPresentInStruct() {

        Schema schema = SchemaBuilder.struct().name("com.example.Person")
                .field("firstName", Schema.STRING_SCHEMA)
                .field("age", Schema.INT32_SCHEMA)
                .field("threshold", Schema.OPTIONAL_FLOAT64_SCHEMA).build();

        Struct struct = new Struct(schema).put("firstName", "Alex").put("age", 30);

        SinkRecord sinkRecord = new SinkRecord("sometopic", 1, null, null, schema, struct, 1);
        new StructFieldsStringKeyBuilder(new String[]{"threshold"}).build(sinkRecord);

    }

    @Test
    public void createTheKeyBasedOnASingleFieldInTheStruct() {
        Schema schema = SchemaBuilder.struct().name("com.example.Person")
                .field("firstName", Schema.STRING_SCHEMA)
                .field("age", Schema.INT32_SCHEMA)
                .field("threshold", Schema.OPTIONAL_FLOAT64_SCHEMA).build();

        Struct struct = new Struct(schema).put("firstName", "Alex").put("age", 30);

        SinkRecord sinkRecord = new SinkRecord("sometopic", 1, null, null, schema, struct, 1);
        assertEquals(new StructFieldsStringKeyBuilder(new String[]{"firstName"}).build(sinkRecord), "Alex");

    }

    @Test
    public void createTheKeyFromMultipleFieldsInTheStruct() {

        Schema schema = SchemaBuilder.struct().name("com.example.Person")
                .field("firstName", Schema.STRING_SCHEMA)
                .field("age", Schema.INT32_SCHEMA)
                .field("threshold", Schema.OPTIONAL_FLOAT64_SCHEMA).build();

        Struct struct = new Struct(schema).put("firstName", "Alex").put("age", 30);

        SinkRecord sinkRecord = new SinkRecord("sometopic", 1, null, null, schema, struct, 1);
        assertEquals(new StructFieldsStringKeyBuilder(new String[]{"firstName", "age"}).build(sinkRecord), "Alex.30");
    }
}
